store/tikv/region_cache.go

```diff
@@ -46,56 +48,159 @@ var (
 	tikvRegionCacheCounterWithGetStoreError         = metrics.TiKVRegionCacheCounter.WithLabelValues("get_store", "err")
 )
 
-// CachedRegion encapsulates {Region, TTL}
-type CachedRegion struct {
-	region     *Region
-	lastAccess int64
+const (
+	updated  int32 = iota // region is updated and no need to reload.
+	needSync              //  need sync new region info.
+)
+
+// Region presents kv region
+type Region struct {
+	meta         *metapb.Region // immutable after fetched from pd
+	stores       []*Store       // stores in this region
+	workStoreIdx int32          // point to current work peer in meta.Peers and work store in stores(same idx)
+	syncStoreIdx int32          // point to init peer index and need reload if meet syncStoreIdx again
+	syncFlag     int32          // mark region need be sync in next turn
+	lastAccess   int64          // last region access time, see checkRegionCacheTTL
+}
+
+func (r *Region) initStores(c *RegionCache) {
+	for _, p := range r.meta.Peers {
+		c.storeMu.RLock()
+		store, exists := c.storeMu.stores[p.StoreId]
+		c.storeMu.RUnlock()
+		if !exists {
+			store = c.getStoreByStoreID(p.StoreId)
+		}
+		r.stores = append(r.stores, store)
+	}
+}
+
+func (r *Region) checkRegionCacheTTL(ts int64) bool {
+retry:
+	lastAccess := atomic.LoadInt64(&r.lastAccess)
+	if ts-lastAccess > rcDefaultRegionCacheTTLSec {
+		return false
+	}
+	if !atomic.CompareAndSwapInt64(&r.lastAccess, lastAccess, ts) {
+		goto retry
+	}
+	return true
+}
+
+// invalidate invalidates a region, next time it will got null result.
+func (r *Region) invalidate() {
+	atomic.StoreInt64(&r.lastAccess, invalidatedLastAccessTime)
+}
+
+// scheduleReload schedules reload region request in next LocateKey.
+func (r *Region) scheduleReload() {
+retry:
+	oldValue := atomic.LoadInt32(&r.syncFlag)
+	if oldValue != updated {
+		return
+	}
+	if !atomic.CompareAndSwapInt32(&r.syncFlag, oldValue, needSync) {
+		goto retry
```

Comment:
I think retry is unnecessary. If the value is changed, it means it is marked `needSync` by an other goroutine. There is no need to mark it again.